File: app/src/main/java/com/github/reline/jisho/ui/controllers/HomeController.java
/*
 * Copyright 2016 Nathaniel Reline
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.reline.jisho.ui.controllers;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.github.reline.jisho.Jisho;
import com.github.reline.jisho.R;
import com.github.reline.jisho.models.Word;
import com.github.reline.jisho.presenters.HomePresenter;
import com.github.reline.jisho.ui.controllers.base.BaseController;
import com.github.reline.jisho.ui.controllers.base.Layout;
import com.github.reline.jisho.ui.recyclerview.WordRecyclerViewAdapter;
import com.github.reline.jisho.ui.views.IHomeView;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;

@Layout(R.layout.controller_home)
public class HomeController extends BaseController implements IHomeView {

    private static final String QUERY = "HomeController.QUERY";
    private static final String WORDS = "HomeController.WORDS";

    @Inject
    HomePresenter presenter;

    @BindView(R.id.progress_bar)
    ProgressBar progressBar;

    @BindView(R.id.logo)
    ImageView logo;

    @BindView(R.id.no_match)
    TextView noMatch;

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    private WordRecyclerViewAdapter adapter;
    private String query;

    public HomeController() {
        Jisho.getInjectionComponent().inject(this);
    }

    @Override
    protected void onViewBound(@NonNull View view) {
        presenter.takeView(this);
        setHasOptionsMenu(true);
        setupRecycler(view.getContext());
    }

    private void setupRecycler(Context context) {
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        adapter = new WordRecyclerViewAdapter();
        recyclerView.setAdapter(adapter);
    }

    @Override
    protected void onRestoreViewState(@NonNull View view, @NonNull Bundle savedViewState) {
        query = savedViewState.getString(QUERY);
        ArrayList<Word> words = savedViewState.getParcelableArrayList(WORDS);
        if (words != null) {
            updateView(words);
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.search_menu, menu);

        MenuItem searchItem = menu.findItem(R.id.action_search);

        SearchView searchView = (SearchView) searchItem.getActionView();
        final Activity activity = getActivity();
        searchView.setQuery(query, false); // restore the query
        searchView.setQueryHint(activity != null ? activity.getString(R.string.search) : null);
        searchView.setMaxWidth(Integer.MAX_VALUE); // allow search view to match the width of the toolbar
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override public boolean onQueryTextChange(String newText) {
                query = newText;
                return true;
            }

            @Override
            public boolean onQueryTextSubmit(String query) {
                progressBar.setVisibility(View.VISIBLE);
                presenter.search(query);
                // hide the keyboard
                View currentFocus = activity != null ? activity.getCurrentFocus() : null;
                if (currentFocus != null) {
                    ((InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE))
                            .hideSoftInputFromWindow(currentFocus.getWindowToken(),
                                    InputMethodManager.HIDE_NOT_ALWAYS);
                }
                return true;
            }
        });
    }

    @Override
    protected void onSaveViewState(@NonNull View view, @NonNull Bundle outState) {
        outState.putString(QUERY, query);
        outState.putParcelableArrayList(WORDS, adapter.getWordList());
    }

    @Override
    protected void onDestroyView(@NonNull View view) {
        presenter.dropView(this);
        super.onDestroyView(view);
    }

    @Override
    public void updateView(@NonNull ArrayList<Word> results) {
        logo.setVisibility(View.GONE);
        adapter.updateData(results);
        progressBar.setVisibility(View.GONE);
        if (results.isEmpty()) {
            noMatch.setVisibility(View.VISIBLE);
            final Context context = getApplicationContext();
            if (context != null) {
                noMatch.setText(String.format(context.getString(R.string.no_match), query));
            }
        } else {
            noMatch.setVisibility(View.GONE);
        }
    }
}

File: jisho/src/main/java/com/github/reline/jisho/ui/controllers/HomeController.kt
/*
 * Copyright 2017 Nathaniel Reline
 *
 * This work is licensed under the Creative Commons Attribution-ShareAlike 4.0 International License.
 * To view a copy of this license, visit http://creativecommons.org/licenses/by-sa/4.0/ or
 * send a letter to Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
 */

package com.github.reline.jisho.ui.controllers

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import com.github.reline.jisho.Jisho
import com.github.reline.jisho.R
import com.github.reline.jisho.models.Word
import com.github.reline.jisho.presenters.HomePresenter
import com.github.reline.jisho.ui.controllers.base.BaseController
import com.github.reline.jisho.ui.controllers.base.Layout
import com.github.reline.jisho.ui.recyclerview.WordRecyclerViewAdapter
import com.github.reline.jisho.ui.views.IHomeView
import kotlinx.android.synthetic.main.controller_home.view.*
import java.util.*
import javax.inject.Inject

@Layout(R.layout.controller_home)
class HomeController : BaseController(), IHomeView {

    @Inject
    internal lateinit var presenter: HomePresenter

    private lateinit var adapter: WordRecyclerViewAdapter
    private var query: String? = null

    init {
        Jisho.getInjectionComponent().inject(this)
    }

    override fun onViewBound(view: View) {
        presenter.takeView(this)
        setHasOptionsMenu(true)
        view.homeControllerRecyclerView.layoutManager = LinearLayoutManager(applicationContext)
        adapter = WordRecyclerViewAdapter()
        view.homeControllerRecyclerView.adapter = adapter
    }

    override fun onRestoreViewState(view: View, savedViewState: Bundle) {
        query = savedViewState.getString(QUERY)
        val words = savedViewState.getParcelableArrayList<Word>(WORDS)
        if (words != null) {
            updateView(words)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.search_menu, menu)

        val searchItem = menu.findItem(R.id.action_search)

        val searchView = searchItem.actionView as SearchView
        searchView.setQuery(query, false) // restore the query
        searchView.queryHint = getString(R.string.search)
        searchView.maxWidth = Integer.MAX_VALUE // allow search view to match the width of the toolbar
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String): Boolean {
                query = newText
                return true
            }

            override fun onQueryTextSubmit(query: String): Boolean {
                view?.homeControllerProgressBar?.visibility = View.VISIBLE
                presenter.search(query)
                hideKeyboard()
                return true
            }
        })
    }

    override fun onSaveViewState(view: View, outState: Bundle) {
        outState.putString(QUERY, query)
        outState.putParcelableArrayList(WORDS, adapter.wordList)
    }

    override fun onDestroyView(view: View) {
        presenter.dropView(this)
        super.onDestroyView(view)
    }

    override fun updateView(results: ArrayList<Word>) {
        view?.apply {
            homeControllerLogoTextView.visibility = View.GONE
            adapter.updateData(results)
            homeControllerProgressBar.visibility = View.GONE
            if (results.isEmpty()) {
                homeControllerNoMatchTextView.visibility = View.VISIBLE
                homeControllerNoMatchTextView.text = getString(R.string.no_match)?.format(query)
            } else {
                homeControllerNoMatchTextView.visibility = View.GONE
            }
        }
    }

    companion object {
        private const val QUERY = "HomeController.QUERY"
        private const val WORDS = "HomeController.WORDS"
    }
}
